### PR TITLE
SONARJAVA-1840 bounds of type var should be used for inference

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 1]
```diff
@@ -71,13 +78,22 @@ TypeSubstitution inferTypeSubstitution(MethodJavaSymbol method, List<JavaType> f
 
   private TypeSubstitution inferTypeSubstitution(MethodJavaSymbol method, TypeSubstitution substitution, JavaType formalType, JavaType argumentType,
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: CRITICAL The Cyclomatic Complexity of this method "inferTypeSubstitution" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* CRITICAL 2 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [TypeSubstitutionSolver.java#L158](https://github.com/SonarSource/sonar-java/blob/2240aa8ff637e9b57186d1f5948a7ac01dcc432e/java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java#L158): The Cyclomatic Complexity of this method "applySubstitution" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. MAJOR [TypeSubstitutionSolver.java#L158](https://github.com/SonarSource/sonar-java/blob/2240aa8ff637e9b57186d1f5948a7ac01dcc432e/java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java#L158): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* CRITICAL 2 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [TypeSubstitutionSolver.java#L158](https://github.com/SonarSource/sonar-java/blob/49eabe6ade2a3351d706522552fc8b6ea3a1864e/java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java#L158): The Cyclomatic Complexity of this method "applySubstitution" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. MAJOR [TypeSubstitutionSolver.java#L158](https://github.com/SonarSource/sonar-java/blob/49eabe6ade2a3351d706522552fc8b6ea3a1864e/java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java#L158): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)

File: java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java [hunk 2]
```diff
@@ -1205,4 +1204,10 @@ private JavaType getRSubstitution(Result result, String symbolName) {
     return typeSubstitution.substitutedTypes().get(1);
   }
 
+  @Test
+  public void infer_from_type_variable_bounds() throws Exception {
+    Result result = Result.createFor("TypeVarBoundsInference");
+    assertThat(result.symbol("write1").usages()).hasSize(1);
+    assertThat(result.symbol("write2").usages()).hasSize(1);
```
Comment (m-g-sonar) [inline java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java]: Add a test on the parameterized method using enums from file `TypeVarBoundsInference.java`, as you used it in your resolution.

File: java-frontend/src/test/files/sym/TypeVarBoundsInference.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
+class Foo {
+  public void foo() {
+    bar(s -> write1(get(s))); // write1 unresolved
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/sym/TypeVarBoundsInference.java]: comments should be removed. We resolve both methods now.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* CRITICAL 2 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [TypeSubstitutionSolver.java#L158](https://github.com/SonarSource/sonar-java/blob/448825b2e0994b96abb4f06533c9a2b1b203c2d0/java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java#L158): The Cyclomatic Complexity of this method "applySubstitution" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. MAJOR [TypeSubstitutionSolver.java#L158](https://github.com/SonarSource/sonar-java/blob/448825b2e0994b96abb4f06533c9a2b1b203c2d0/java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java#L158): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)

File: java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java (hunk 2 above)
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java]: fixed

File: java-frontend/src/test/files/sym/TypeVarBoundsInference.java (hunk 3 above)
Comment (benzonico) [inline java-frontend/src/test/files/sym/TypeVarBoundsInference.java]: fixed

**Pull request merged**